### PR TITLE
ARROW-10557: [C++] Add scalar string slicing/substring extract kernel

### PR DESCRIPTION
**Pull request by maartenbreddels**
Needs a rebase after https://github.com/apache/arrow/pull/8621 is merged

I totally agree with https://github.com/python/cpython/blob/c9bc290dd6e3994a4ead2a224178bcba86f0c0e4/Objects/sliceobject.c#L252 

This was tricky to get right, the main difficulty is in manually dealing with reverse iterators. Therefore I put on extra guardrails by having the Python unittests cover a lot of cases. All edge cases detected by this are translated to the C++ unittest suite, so we could reduce them to reduce pytest execution cost (I added 1 second).

Slicing is based on Python, `[start, stop)` inclusive/exclusive semantics, where an index refers to a codeunit (like Python apparently, badly documented), and negative indices start counting from the right. `step != 0` is supported, like Python. 

The only thing we cannot support easily, are things like reversing a string, since in Python one can do `s[::-1]` or `s[-1::-1]`, but we don't support empty values with the Option machinery (we model this as an c-`int64`). To mimic this, we can do `pc.utf8_slice_codeunits(ar, start=-1, end=-sys.maxsize, step=-1)` (i.e. a very large negative value). 

For instance, libraries such as Pandas and Vaex can do sth like that, confirmed to be working by modifying the unittest like this:
```python
import sys
@pytest.mark.parametrize('start', list(range(-6, 6)) + [None])
@pytest.mark.parametrize('stop', list(range(-6, 6)) + [None])
@pytest.mark.parametrize('step', [-3, -2, -1, 1, 2, 3])
def test_slice_compatibility(start,stop, step):
    input = pa.array(["", "𝑓", "𝑓ö", "𝑓öõ", "𝑓öõḍ", "𝑓öõḍš"])
    expected = pa.array([k.as_py()[start:stop:step] for k in input])
    if start is None:
        start = -sys.maxsize if step > 0 else sys.maxsize
    if stop is None:
        stop = sys.maxsize if step > 0 else -sys.maxsize
    result = pc.utf8_slice_codeunits(input, start=start, stop=stop, step=step)
    assert expected.equals(result)
```

So libraries using this can implement the full Python behavior with this workaround.

### CODE REVIEW

**Comment by maartenbreddels** [comment]
@pitrou this is ready for review

**Comment by pitrou** [comment]
> All edge cases detected by this are translated to the C++ unittest suite, so we could reduce them to reduce pytest execution cost (I added 1 second).

You could instead use regular for loops instead of the pytest parametrization "magic". It would probably remove a lot of overhead.

**Comment by pitrou** [comment]
I'll think I'll do a full review once the (lr)trim PR is merged, it will be easier.

**Comment by nealrichardson** [comment]
👋 checking back in here. The PR on which this was based has merged, so this can be rebased now

**Comment by maartenbreddels** [comment]
@pitrou this is ready for review, the failure seems unrelated (minio)

Sorry for taking so long to get back at this, I hope we can get this, and my other open string-PRs sorted out soon!

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 1]
```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
...
+    "Slice string ",
+    ("For each string in `strings`, slice into a substring defined by\n"
+     "`start`, `stop`, `step`) as given by `SliceOptions` where `start` is inclusive\n"
+     "and `stop` is exclusive and are measured in codeunits. If step is negative, the\n"
+     "string will be advanced in reversed order. A `step` of zero is considered an\n"
+     "error.\n"
+     "Null inputs emit null."),
+    {"strings"}, "SliceOptions");
+
+void AddSlice(FunctionRegistry* registry) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:604]: Is there a reason we don't also add a `ascii_slice` for binary and string?

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 2]
```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
...
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
+    const uint8_t* begin = input;
+    const uint8_t* end = input + input_string_ncodeunits;
+    const uint8_t* begin_sliced = begin;
+    const uint8_t* end_sliced = end;
+    SliceOptions& options = this->options;
+
+    if (options.step >= 1) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:476]: The combinatory explosion in `if` branches seems a bit gratuitous? Abstractly it seems you should be able to do:
```c++
// Compute begin_sliced
if (options.start >= 0) {
  PROPAGATE_FALSE(
            arrow::util::UTF8AdvanceCodepoints(begin, end, &begin_sliced, options.start));
} else {
  PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(
            begin, end, &begin_sliced, -options.start));
}
// Compute end_sliced
if (options.stop >= 0) {
  PROPAGATE_FALSE(
            arrow::util::UTF8AdvanceCodepoints(begin, end, &end_sliced, options.stop));
} else {
  PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(
            begin, end, &end_sliced, -options.stop));
}
if (options.step > 0) {
  if (end_sliced < begin_sliced) {
    // empty slice
    *output_written = 0;
    return true;
  }
  // ...
}
else {
  DCHECK_LT(options.step, 0);
  if (end_sliced > begin_sliced) {
    // empty slice
    *output_written = 0;
    return true;
  }
  // ...
}
```

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 3]
```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
...
+  using offset_type = typename Base::offset_type;
+  using Base::Base;
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
+    const uint8_t* begin = input;
+    const uint8_t* end = input + input_string_ncodeunits;
+    const uint8_t* begin_sliced = begin;
+    const uint8_t* end_sliced = end;
+    SliceOptions& options = this->options;
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:474]: `const&`

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 4]
```diff
@@ -462,6 +462,107 @@ TYPED_TEST(TestStringKernels, TrimUTF8) {
...
+// produce test data with e.g.:
+// repr([k[-3:1] for k in ["", "𝑓", "𝑓ö", "𝑓öõ", "𝑓öõḍ", "𝑓öõḍš"]]).replace("'", '"')
+
+#ifdef ARROW_WITH_UTF8PROC
+TYPED_TEST(TestStringKernels, SliceCodeunitsBasic) {
+  SliceOptions options{2, 4};
+  this->CheckUnary("utf8_slice_codeunits", R"(["foo", "fo", null, "foo bar"])",
+                   this->type(), R"(["o", "", null, "o "])", &options);
+  SliceOptions options_2{2, 3};
+  // ensure we slice in codeunits, not graphmemes
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:474]: "graphemes"

File: docs/source/cpp/compute.rst [hunk 5]
```diff
@@ -553,6 +553,25 @@ when a positive ``max_splits`` is given.
...
 
+Slicing
+~~~~~~~
+
+These function transform each sequence of the array to a subsequence, following Python slicing semantics.
+
++--------------------------+------------+-------------------------+-------------------+----------------------------------+---------+
+| Function name            | Arity      | Input types             | Output type       | Options class                    | Notes   |
++==========================+============+=========================+===================+==================================+=========+
+| utf8_slice_codepoints    | Unary      | String-like             | String-like         | :struct:`SliceOptions`         | \(1)    |
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:564]: There is a slight alignment bug in your reST markup here.

File: docs/source/cpp/compute.rst [hunk 6]
```diff
@@ -553,6 +553,25 @@ when a positive ``max_splits`` is given.
...
++--------------------------+------------+-------------------------+-------------------+----------------------------------+---------+
+
+
+* \(1) Slice string into a substring defined by (`start`, `stop`, `step`) as
+given by `SliceOptions` where `start` is inclusive and `stop` is exclusive and
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:569]: Must probably indent these lines to be part of the bullet item (you can click "view file" to get a reST rendering in the Github UI).

File: docs/source/cpp/compute.rst [hunk 7]
```diff
@@ -553,6 +553,25 @@ when a positive ``max_splits`` is given.
...
+These function transform each sequence of the array to a subsequence, following Python slicing semantics.
+
++--------------------------+------------+-------------------------+-------------------+----------------------------------+---------+
+| Function name            | Arity      | Input types             | Output type       | Options class                    | Notes   |
++==========================+============+=========================+===================+==================================+=========+
+| utf8_slice_codepoints    | Unary      | String-like             | String-like         | :struct:`SliceOptions`         | \(1)    |
++--------------------------+------------+-------------------------+-------------------+----------------------------------+---------+
+
+
+* \(1) Slice string into a substring defined by (`start`, `stop`, `step`) as
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:568]: Need double-backticks in reST:
```reST
* \(1) Slice string into a substring defined by (``start``, ``stop``, ``step``) as
```

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 8]
```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
...
+  } while (0)
+
+template <typename Type>
+struct SliceCodeunits : SliceBase<Type, SliceCodeunits<Type>> {
+  using Base = SliceBase<Type, SliceCodeunits<Type>>;
+  using offset_type = typename Base::offset_type;
+  using Base::Base;
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:469]: Note that most of this function is really identical regardless of the offset type. So instead you could define a `Utf8SliceBase` base class:
```c++
struct Utf8SliceBase {
  bool Transform(const uint8_t* input, int64_t input_string_ncodeunits,
                 uint8_t* output, int64_t* output_written) {
    // ...
  }
};
```
... and inherit from that mixin class in concrete classes:
```c++
template <typename Type>
struct SliceCodeunits : SliceBase<Type, SliceCodeunits<Type>>, Utf8SliceBase {
  using Base = SliceBase<Type, SliceCodeunits<Type>>;
  using offset_type = typename Base::offset_type;
  using Base::Base;

  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
                 uint8_t* output, offset_type* output_written) {
    int64_t output_written_64;
    bool res = Utf8SliceBase::Transform(input, input_string_ncodeunits, output, &output_written_64);
    *output_written = static_cast<offset_type>(output_written_64);
    return res;
  }
```

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 9]
```diff
@@ -462,6 +462,107 @@ TYPED_TEST(TestStringKernels, TrimUTF8) {
...
+  this->CheckUnary("utf8_slice_codeunits", R"(["", "𝑓", "𝑓ö", "𝑓öõ", "𝑓öõḍ", "𝑓öõḍš"])",
+                   this->type(), R"(["", "𝑓", "𝑓", "𝑓õ", "𝑓õ", "öḍ"])", &options_step);
+  SliceOptions options_step_neg{-1, 1, -2};
+  this->CheckUnary("utf8_slice_codeunits", R"(["", "𝑓", "𝑓ö", "𝑓öõ", "𝑓öõḍ", "𝑓öõḍš"])",
+                   this->type(), R"(["", "", "", "õ", "ḍ", "šõ"])", &options_step_neg);
+  options_step_neg.stop = 0;
+  this->CheckUnary("utf8_slice_codeunits", R"(["", "𝑓", "𝑓ö", "𝑓öõ", "𝑓öõḍ", "𝑓öõḍš"])",
+                   this->type(), R"(["", "", "ö", "õ", "ḍö", "šõ"])", &options_step_neg);
+}
+
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:563]: Nice tests, thank you!

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 10]
```diff
@@ -462,6 +462,107 @@ TYPED_TEST(TestStringKernels, TrimUTF8) {
...
+  SliceOptions options_edgecase_1{-3, 1};
+  this->CheckUnary("utf8_slice_codeunits", R"(["𝑓öõḍš"])", this->type(), R"([""])",
+                   &options_edgecase_1);
+
+  // this is a safeguard agains an optimization path possible, but actually a tricky case
+  SliceOptions options_edgecase_2{-6, -2};
+  this->CheckUnary("utf8_slice_codeunits", R"(["𝑓öõḍš"])", this->type(), R"(["𝑓öõ"])",
+                   &options_edgecase_2);
+}
+
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:501]: By the way, what happens if I call `utf8_slice_codeunits` without an options instance? Should you add a test for that?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you very much @maartenbreddels . Here are some comments.

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 1 above)
Comment (maartenbreddels) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:632]: No, just focussing on the harder problems first.

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 2 above)
Comment (maartenbreddels) [reply cpp/src/arrow/compute/kernels/scalar_string.cc]: The idea is that you don't start counting from the beginning or end twice when possible, e.g. if you do slice from -6 to -5, it's faster to not start counting code points from the end twice. It increases code complexity, that's the payoff.

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 8 above)
Comment (maartenbreddels) [reply cpp/src/arrow/compute/kernels/scalar_string.cc]: Yes, good catch. Put it in a function now.

**Comment by maartenbreddels** [comment]
Failure seems unrelated: `npm WARN tar ENOSPC: no space left on device, write`

**Comment by nealrichardson** [comment]
@pitrou @maartenbreddels is this ready to merge? Trying to mop up issues for 4.0

**Comment by pitrou** [comment]
@nealrichardson I need to review this again, sorry.

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 11]
```diff
@@ -447,6 +447,197 @@ void AddMatchSubstring(FunctionRegistry* registry) {
...
+        *output_written = 0;
+        return true;
+      }
+    } else {
+      // start counting from the right
+      PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(begin, end, &begin_sliced,
+                                                                -options.start));
+      if (options.stop > 0) {
+        // continue counting from the left, we cannot start from begin_sliced because we
+        // don't know how many codepoints are between begin and begin_sliced
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:515]: Hmm... for `step < 0`, wouldn't it be more logical to first compute `end_sliced`? Presumably, you then can pass `end_sliced` as one of the boundaries for computing `begin_sliced`?

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:515]: Just a question btw, you don't need to act on this if you think it's unnecessary.

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 12]
```diff
@@ -447,6 +447,197 @@ void AddMatchSubstring(FunctionRegistry* registry) {
...
+    }
+    Derived(options).Execute(ctx, batch, out);
+  }
+
+  void Execute(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    Base::Execute(ctx, batch, out);
+  }
+};
+
+#define PROPAGATE_FALSE(expr)         \
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:475]: Can you `#undef` this at the end?

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 13]
```diff
@@ -522,6 +522,113 @@ TYPED_TEST(TestStringKernels, TrimUTF8) {
...
+  SliceOptions options_edgecase_2{-6, -2};
+  this->CheckUnary("utf8_slice_codeunits", R"(["𝑓öõḍš"])", this->type(), R"(["𝑓öõ"])",
+                   &options_edgecase_2);
+
+  auto input = ArrayFromJSON(this->type(), R"(["𝑓öõḍš"])");
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      testing::HasSubstr("Attempted to initialize KernelState from null FunctionOptions"),
+      CallFunction("utf8_slice_codeunits", {input}));
+}
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:566]: Also add a test with `stop = 0`?

**Review by pitrou — CHANGES_REQUESTED** [review]
Just two more questions. Looks good otherwise!

**Comment by pitrou** [comment]
@maartenbreddels Do you want to update this PR?

**Comment by pitrou** [comment]
I'm working on updating this.

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**